components/brave_wallet/browser/cardano/cardano_provider_impl_unittest.cc

```diff
@@ -447,4 +447,314 @@ TEST_F(CardanoProviderImplUnitTest, EnableFails_OnWalletUnlock_TabNotActive) {
   EXPECT_TRUE(error);
 }
 
+// Security Tests - Additional Coverage
+
+TEST_F(CardanoProviderImplUnitTest, Enable_GlobalPermissionDenial) {
+  // Security test: Verify that global permission denial cannot be bypassed
+  CreateWallet();
+  UnlockWallet();
+  auto added_account = AddAccount();
+
+  ON_CALL(*delegate(), IsTabVisible()).WillByDefault([&]() { return true; });
+
+  // Mock global permission denial
+  ON_CALL(*delegate(), IsPermissionDenied(mojom::CoinType::ADA))
+      .WillByDefault([&]() { return true; });
+
+  TestFuture<mojom::CardanoProviderErrorBundlePtr> future;
+  mojo::Remote<mojom::CardanoApi> api;
+  provider()->Enable(api.BindNewPipeAndPassReceiver(), future.GetCallback());
+
+  auto error = future.Take();
+  ASSERT_TRUE(error);
+  EXPECT_EQ(error->code, -3);  // kAPIErrorRefused
+  EXPECT_EQ(error->error_message,
+            l10n_util::GetStringUTF8(IDS_WALLET_USER_REJECTED_REQUEST));
+}
+
+TEST_F(CardanoProviderImplUnitTest, IsEnabled_GlobalPermissionDenial) {
+  // Security test: Verify IsEnabled respects global denial
+  CreateWallet();
+  UnlockWallet();
+  auto added_account = AddAccount();
+
+  ON_CALL(*delegate(), IsTabVisible()).WillByDefault([&]() { return true; });
+
+  // Mock global permission denial
+  ON_CALL(*delegate(), IsPermissionDenied(mojom::CoinType::ADA))
+      .WillByDefault([&]() { return true; });
+
+  TestFuture<bool> future;
+  provider()->IsEnabled(future.GetCallback());
+  EXPECT_FALSE(future.Take());
+}
+
+TEST_F(CardanoProviderImplUnitTest, Enable_OriginValidation) {
+  // Security test: Verify origin is obtained from delegate (browser-side)
+  CreateWallet();
+  UnlockWallet();
+  auto added_account = AddAccount();
+
+  const url::Origin expected_origin =
+      url::Origin::Create(GURL("https://example.com"));
+
+  // Expect GetOrigin to be called
+  EXPECT_CALL(*delegate(), GetOrigin())
+      .WillOnce(testing::Return(expected_origin));
+
+  ON_CALL(*delegate(), IsTabVisible()).WillByDefault([&]() { return true; });
+
+  ON_CALL(*delegate(), GetAllowedAccounts(_, _))
+      .WillByDefault(
+          [&](mojom::CoinType coin, const std::vector<std::string>& accounts) {
+            return std::vector<std::string>(
+                {added_account->account_id->unique_key});
+          });
+
+  TestFuture<mojom::CardanoProviderErrorBundlePtr> future;
+  mojo::Remote<mojom::CardanoApi> api;
+  provider()->Enable(api.BindNewPipeAndPassReceiver(), future.GetCallback());
+
+  auto error = future.Take();
+  EXPECT_FALSE(error);
+}
+
+TEST_F(CardanoProviderImplUnitTest, Enable_UnlockRetainsCapturedOrigin) {
```

Comment:
this isn't really needed since the response calls back to the origin via the promise, but is kind of useful for the defense in depth check